app/services/authentication/strategy/infer_service.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+module Authentication
+  module Strategy
+    class InferService < ThreeScale::Patterns::Service
+
+      def initialize(params, site_account, admin_domain: false)
+        @params = params
+        @site_account = site_account
+        @admin_domain = admin_domain
+      end
+
+      def call
+        type = infer_type
+
+        return build(:token) if impersonating?(type)
+        return build(:provider_oauth2) if sso_enforced?
```

Comment:
```suggestion
        return :null unless type
        return build(:provider_oauth2) if sso_enforced? && !impersonating?(type)
```

I needed some time to figure out (hopefully correctly) that we don't want to enforce sso when we are impersonating. So in this way, to me it is more readable. Otherwise I was wondering why do we have `:token` downstairs and here we handle differently. So unless you like better the original, here's my suggested change.

wrt `return :null`, the idea is that we should not got to the rescue in `app/lib/authentication/strategy.rb`, instead the Null strategy is business as usual. I assume we can remove the rescue there or add logging because somthing needs to be very broken if we reach there.